[GeoMechanicsApplication] Modify the input parameters for the elastic matrix calculation

avdg81 opened the pull request:
**📝 Description**

When calculating the elastic matrix, it is natural to provide Young's modulus $`E`$ and Poisson's ratio $`\nu`$. These new parameters replace the set of three coefficients that had to be supplied previously.

**🆕 Changelog**

Other changes include:
- Renamed the member function that calculates the elastic matrix.
- Added a member function that returns the number of normal components in the stress/strain vectors.
- The incremental linear elastic constitutive law sets the off-diagonal entries as well as the shear-related entries to zero upon request.

avdg81 committed:
Provide $`E`$ and $`\nu`$ to compute the elastic matrix

Young's modulus and Poisson's ratio are natural input parameters for a function that calculates the elastic matrix.

Other changes include:
- Renamed the member that calculates the elastic matrix.
- Added `[[nodiscard]]` attributes to most member functions of abstract class `ConstitutiveLawDimension`.
- The incremental linear elastic constitutive law sets entries of the elastic matrix to zero when option "diagonal entries only and no shear" has been set. This is now done by creating a new matrix and only filling the relevant diagonal entries. This needs to be optimized.

avdg81 committed:
Added a query for the number of normal components

All constitutive dimensions must now implement a query that returns the normal of normal components.

avdg81 committed:
Avoid copying a matrix

Set the off-diagonal entries as well as any shear entries to zero when requested. No longer create a new matrix, copy the relevant diagonal entries, and copy this matrix to the target matrix.

avdg81 committed:
Added a few debug checks

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/incremental_linear_elastic_law.cpp:
```diff
@@ -14,6 +14,44 @@
+
+void SetEntriesAboveDiagonalToZero(Matrix& rMatrix)
+{
+    KRATOS_DEBUG_ERROR_IF(rMatrix.size1() != rMatrix.size2()) << "Matrix must be square\n";
```
( Nagging! ) Maybe this condition should reflect in the function name e.g. SetEntriesAboveDiagonalOfSquareMatrixToZero 
and simillar for the below and diagonal version.
However, I think that the algorithm works for non square matrices too, The error could probably be omitted.

avdg81 replied:
I agree with you that this function (and the two companion functions) should also work for non-square matrices. So I have removed the error check and kept the function names as-is.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/incremental_linear_elastic_law.cpp:
```diff
@@ -101,15 +139,14 @@ void GeoIncrementalLinearElasticLaw::CalculateElasticMatrix(Matrix& C, Constitut
-    const double c3 = this->GetConsiderDiagonalEntriesOnlyAndNoShear() ? 0.0 : (0.5 - NU) * c0;
-
-    C = mpConstitutiveDimension->FillConstitutiveMatrix(c1, c2, c3);
+    if (this->GetConsiderDiagonalEntriesOnlyAndNoShear()) {
+        SetEntriesAboveDiagonalToZero(C);
+        SetEntriesBelowDiagonalToZero(C);
+        SetShearEntriesToZero(C, mpConstitutiveDimension->GetNumberOfNormalComponents());
+    }
```
This really is a nicer implementation.

avdg81 replied:
Thanks.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/plane_strain.cpp:
```diff
@@ -18,8 +18,13 @@
+    const auto c0 = YoungsModulus / ((1.0 + PoissonsRatio) * (1.0 - 2.0 * PoissonsRatio));
+    const auto c1 = (1.0 - PoissonsRatio) * c0;
+    const auto c2 = c0 * PoissonsRatio;
+    const auto c3 = (0.5 - PoissonsRatio) * c0;
```
For improved readability, please be consistent with c0 either always first or always last.

avdg81 replied:
I have put coefficient `c0` consistently at the back side now.

WPK4FEM reviewed (commented):
I think most is an improvement. Some nagging about readability and necessity of error condition.

Wijtze Pieter

avdg81 committed:
Processed Wijtze Pieter's review suggestions

- Removed the (redundant) error checks for square matrices, since the functions also work properly for non-square matrices.
- When calculating the matrix coefficients, consistently put `c0` at the back.

rfaasse reviewed (approved):
Nice implementation, reads a lot clearer than before! From my side this is ready to go!
